File: lib/services/internet_speed_test_service.dart
import 'dart:async';

import 'package:http/http.dart' as http;

class InternetSpeedTestResult {
  final double downloadMbps;
  final int latencyMs;
  final int bytesTransferred;
  final Duration transferDuration;
  final DateTime measuredAt;

  const InternetSpeedTestResult({
    required this.downloadMbps,
    required this.latencyMs,
    required this.bytesTransferred,
    required this.transferDuration,
    required this.measuredAt,
  });
}

class InternetSpeedTestService {
  InternetSpeedTestService._();

  static final InternetSpeedTestService instance =
      InternetSpeedTestService._();

  static const String _userAgent = 'TV FULL Internet Test/1.0';
  static final Uri _downloadEndpoint =
      Uri.parse('https://speed.cloudflare.com/__down');

  final http.Client _client = http.Client();

  Future<InternetSpeedTestResult> run({
    Duration requestTimeout = const Duration(seconds: 15),
  }) async {
    final latencySamples = <int>[];
    for (var i = 0; i < 3; i++) {
      final sample = await _download(
        1024,
        timeout: const Duration(seconds: 6),
        sampleId: 'latency-$i',
      );
      latencySamples.add(sample.elapsed.inMilliseconds.clamp(1, 60000));
    }
    latencySamples.sort();
    final latencyMs = latencySamples[latencySamples.length ~/ 2];

    // Calentamiento corto para evitar que DNS/TLS domine la medición grande.
    await _download(
      256 * 1024,
      timeout: requestTimeout,
      sampleId: 'warmup',
    );

    // Dos tamaños progresivos permiten que conexiones rápidas tengan tiempo de
    // alcanzar velocidad sostenida sin consumir una cantidad excesiva de datos.
    final samples = <_DownloadSample>[];
    for (final bytes in const [2 * 1024 * 1024, 8 * 1024 * 1024]) {
      samples.add(
        await _download(
          bytes,
          timeout: requestTimeout,
          sampleId: 'download-$bytes',
        ),
      );
    }

    final usable = samples.where((sample) => sample.bytes > 0).toList();
    if (usable.isEmpty) {
      throw Exception('No se recibieron datos suficientes para medir la velocidad.');
    }

    // Tomamos la mejor muestra sostenida. La muestra pequeña sirve para redes
    // lentas; la de 8 MB evita subestimar enlaces rápidos por el costo inicial.
    var bestMbps = 0.0;
    var totalBytes = 0;
    var totalDuration = Duration.zero;
    for (final sample in usable) {
      final seconds = sample.elapsed.inMicroseconds / 1000000.0;
      if (seconds <= 0) continue;
      final mbps = (sample.bytes * 8) / seconds / 1000000.0;
      if (mbps > bestMbps) bestMbps = mbps;
      totalBytes += sample.bytes;
      totalDuration += sample.elapsed;
    }

    if (bestMbps <= 0) {
      throw Exception('No se pudo calcular una velocidad de descarga válida.');
    }

    return InternetSpeedTestResult(
      downloadMbps: bestMbps,
      latencyMs: latencyMs,
      bytesTransferred: totalBytes,
      transferDuration: totalDuration,
      measuredAt: DateTime.now(),
    );
  }

  Future<_DownloadSample> _download(
    int bytes, {
    required Duration timeout,
    required String sampleId,
  }) async {
    final uri = _downloadEndpoint.replace(
      queryParameters: <String, String>{
        'bytes': '$bytes',
        // Evita que intermediarios reutilicen una respuesta anterior.
        'tvfull': '${DateTime.now().microsecondsSinceEpoch}-$sampleId',
      },
    );

    final stopwatch = Stopwatch()..start();
    final request = http.Request('GET', uri)
      ..headers.addAll(const {
        'User-Agent': _userAgent,
        'Accept': 'application/octet-stream,*/*',
        'Cache-Control': 'no-cache',
      });

    final response = await _client.send(request).timeout(timeout);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      stopwatch.stop();
      throw Exception(
        'El servidor de prueba respondió HTTP ${response.statusCode}.',
      );
    }

    var received = 0;
    await for (final chunk in response.stream.timeout(timeout)) {
      received += chunk.length;
    }
    stopwatch.stop();

    return _DownloadSample(bytes: received, elapsed: stopwatch.elapsed);
  }
}

class _DownloadSample {
  final int bytes;
  final Duration elapsed;

  const _DownloadSample({required this.bytes, required this.elapsed});
}

File: native-tv-complete/app/src/main/java/com/tvfull/pro/InternetSpeedTestService.kt
package com.tvfull.pro

import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.max

data class InternetSpeedTestResult(
    val downloadMbps: Double,
    val latencyMs: Long,
    val bytesTransferred: Long,
    val transferDurationMs: Long
)

object InternetSpeedTestService {
    private const val ENDPOINT = "https://speed.cloudflare.com/__down"
    private const val UA = "TV FULL Internet Test/1.0"

    fun run(): InternetSpeedTestResult {
        val latencies = mutableListOf<Long>()
        repeat(3) { index ->
            val sample = download(1024, 6_000, "latency-$index")
            latencies += sample.elapsedMs.coerceIn(1, 60_000)
        }
        latencies.sort()
        val latency = latencies[latencies.size / 2]

        download(256 * 1024, 12_000, "warmup")

        val samples = listOf(
            download(2 * 1024 * 1024, 15_000, "download-2m"),
            download(8 * 1024 * 1024, 15_000, "download-8m")
        ).filter { it.bytes > 0 && it.elapsedMs > 0 }

        if (samples.isEmpty()) throw IllegalStateException("No se recibieron datos suficientes")

        var best = 0.0
        var totalBytes = 0L
        var totalDuration = 0L
        for (sample in samples) {
            val seconds = sample.elapsedMs / 1000.0
            val mbps = (sample.bytes * 8.0) / seconds / 1_000_000.0
            best = max(best, mbps)
            totalBytes += sample.bytes
            totalDuration += sample.elapsedMs
        }
        if (best <= 0.0) throw IllegalStateException("No se pudo calcular la velocidad")

        return InternetSpeedTestResult(best, latency, totalBytes, totalDuration)
    }

    private fun download(bytes: Int, timeoutMs: Int, sampleId: String): Sample {
        val url = URL("$ENDPOINT?bytes=$bytes&tvfull=${System.nanoTime()}-$sampleId")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            connectTimeout = timeoutMs.coerceAtMost(8_000)
            readTimeout = timeoutMs
            instanceFollowRedirects = true
            useCaches = false
            setRequestProperty("User-Agent", UA)
            setRequestProperty("Accept", "application/octet-stream,*/*")
            setRequestProperty("Cache-Control", "no-cache")
        }
        val start = System.nanoTime()
        try {
            val status = conn.responseCode
            if (status !in 200..299) throw IllegalStateException("Test HTTP $status")
            var total = 0L
            val buffer = ByteArray(32 * 1024)
            conn.inputStream.use { input ->
                while (true) {
                    val n = input.read(buffer)
                    if (n <= 0) break
                    total += n
                }
            }
            val elapsed = ((System.nanoTime() - start) / 1_000_000L).coerceAtLeast(1L)
            return Sample(total, elapsed)
        } finally {
            conn.disconnect()
        }
    }

    private data class Sample(val bytes: Long, val elapsedMs: Long)
}
